src/ui/public/kfetch/kfetch.js

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import 'isomorphic-fetch';
+import url from 'url';
+import chrome from '../chrome';
+import { metadata } from '../metadata';
+import { merge } from 'lodash';
+
+class FetchError extends Error {
+  constructor(res, body) {
+    super(res.statusText);
+    this.res = res;
+    this.body = body;
+    Error.captureStackTrace(this, FetchError);
+  }
+}
+
+export function kfetch(fetchOptions, kibanaOptions) {
+  // fetch specific options with defaults
+  const { pathname, query, ...combinedFetchOptions } = merge(
+    {
+      method: 'GET',
+      credentials: 'same-origin',
+      headers: {
+        'Content-Type': 'application/json',
+        'kbn-version': metadata.version,
+      },
+    },
+    fetchOptions,
+  );
+
+  // kibana specific options with defaults
+  const combinedKibanaOptions = {
+    prependBasePath: true,
+    ...kibanaOptions,
+  };
+
+  const fullUrl = url.format({
+    pathname: combinedKibanaOptions.prependBasePath ? chrome.addBasePath(pathname) : pathname,
+    query,
+  });
+
+  const fetching = new Promise(async (resolve, reject) => {
+    const res = await fetch(fullUrl, combinedFetchOptions);
+
+    if (!res.ok) {
+      let body;
+      try {
+        body = await res.json();
+      } catch (err) {
+        // ignore error, may not be able to get body for response that is not ok
+      }
+      reject(new FetchError(res, body));
+    }
+
+    resolve(res.json());
+  });
```

Comment:
We don't need the Promise constructor anymore, do we? Eg. just revert to what we had (unless you prefer this style ofc :) )

Reply:
If you don't mind, I would like to keep it in this style because it makes it just a _bit_ more explicit that a promise is being returned (at least to me!)